### ISSUE TITLE
Unable to install Azure CLI on Windows without admin access to machine

### ISSUE THREAD

**Issue by scrofungulus** [Azure/azure-cli#15381]
**Is your feature request related to a problem? Please describe.**
As the title states, there is not currently a way (or I have not been able to find one) to install the Azure CLI without having admin privileges to the machine. This is fine for personal machines, but employer owned machines are typically highly locked down.

**Describe the solution you'd like**
A sane way to install the cli without admin privileges.

**Comment by bquantump** [comment]
@scrofungulus Thank you for your request. 

A work around is to pip install the CLI: pip install azure-cli

It is recommend you do this through a virtual environment (so it doesn't apply system wide) although not required: https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/

**Comment by david-jarman** [comment]
Was hoping this would unblock me but I ran into an issue with the pip install:
```
error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
----------------------------------------
ERROR: Failed building wheel for cryptography
```

And ofc you need Admin Privileges to install "Microsoft C++ Build Tools"

**Comment by l3ender** [comment]
Is there any standalone executable that can be used instead of the admin install? We are not able to use the CLI in our organization. We have Windows but Python is not installed.

**Comment by isaiahpettingill** [comment]
Python does not typically compile to a native executable. The Azure CLI requires a python interpreter.

**Comment by jiasli** [comment]
As @bquantump pointed out in https://github.com/Azure/azure-cli/issues/15381#issuecomment-702328246, currently installing via `pip` is the only option to install Azure CLI without admin permission. You may also use this approach to run Azure CLI as 64-bit (https://github.com/Azure/azure-cli/issues/18766):

1. Download Python from the official site https://www.python.org/downloads/windows/. **Windows installer (64-bit)** is preferred over **Windows installer (32-bit)**.
2. Install Python which doesn't require admin permission. It will be installed to 
   ```
   C:\Users\<username>\AppData\Local\Programs\Python\Python39\python.exe
   ```
3. (Optional) Upgrade `pip`:
   ```
   python -m pip install --upgrade pip
   ```
4. Run `pip` to install Azure CLI from PyPI https://pypi.org/project/azure-cli/: 
    ```
    pip install azure-cli
    ```

@david-jarman, for error https://github.com/Azure/azure-cli/issues/15381#issuecomment-713155243

```
error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
----------------------------------------
ERROR: Failed building wheel for cryptography
```

This is because you are using an old version of `pip` which can't resolve the pre-compiled library correctly. You may follow the above step 3 to upgrade `pip` first and try again.

**Comment by nrdev88** [comment]
Please fix the installer to not require admin rights. :-) This is a major difference compared to the "classic CLI".

**Comment by jiasli** [comment]
## Workaround

I have provided more detailed instruction on how to install with pip at https://github.com/Azure/azure-cli/issues/20476.

**Comment by paggarwalmsft** [comment]
Hi, is there any update on this issue?

**Comment by jiasli** [comment]
Nope. We'll make sure to update this thread when there is. In the meantime, might I recommend the **Subscribe** button?

![image](https://user-images.githubusercontent.com/4003950/150928883-3ab5b269-0df3-49ac-b2fa-6b3c704c33e1.png)

That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread. ☺️

**Comment by ozbillwang** [comment]
@jiasli 's workaround to install az cli via pip is good enough. 

Thanks.

**Comment by patricekrakow** [comment]
The installation of azure CLI `az` using `pip` did work for me on my _employer-owned machine_. Thanks to @jiasli.

**Comment by sdg002** [comment]
I have been facing this problem at my new employer.  I needed IT intervention to get AZ CLI to get installed.

But, the problem does not stop here.  When I use the command `az upgrade` , it fails once again due to me not having Admin rights on the local workstation.

May I ask what was Microsoft's rationale in chosing  such an installation model?

**Comment by VOVELEE** [comment]
It will be great feature and I think many people are anticipating it

**Comment by JiBo87** [comment]
Im following this. My employer revoked all admin privileges for engineers.

**Comment by ctahok** [comment]
I am waiting for one too

**Comment by ctahok** [comment]
> Im following this. My employer revoked all admin privileges for engineers.

The same is here

**Comment by andyxuan2010** [comment]
The same is here. it would be very handy feature!!!

**Comment by golaat** [comment]
Will this ever be resolved? While is local admin access required?

**Comment by ozbillwang** [comment]
will this be helpful?

https://stackoverflow.com/questions/75377993/how-to-install-az-cli-on-a-locked-up-windows-10-without-admin-rights-behind-a-co/75378210#75378210

**Comment by sdg002** [comment]
> will this be helpful?
> 
> https://stackoverflow.com/questions/75377993/how-to-install-az-cli-on-a-locked-up-windows-10-without-admin-rights-behind-a-co/75378210#75378210

Interesting approach. 
I am divided. Are these many extra steps really needed in the first place ? Shouldn't MS be fixing the obvious problem? Or is MS listening at all?

**Comment by l3ender** [comment]
> will this be helpful?
> 
> https://stackoverflow.com/questions/75377993/how-to-install-az-cli-on-a-locked-up-windows-10-without-admin-rights-behind-a-co/75378210#75378210

We're using this approach as a workaround, but have found it has some issues.

The largest one is that the pip version of az-cli can cause conflicts if you have installed other Azure-related python dependencies. We run into this when using Ansible and the Azure collection (which under the hood uses specific Azure python dependencies). To get around this, we install az-cli in its own virtual environment. This has issues of its own, but it at least provides a path forward.

Secondly, we found the release cycle for az-cli differs than the full tool, so bugs aren't always resolved at the same time.

**Comment by jiasli** [comment]
@l3ender,

> To get around this, we install az-cli in its own virtual environment.

Installing Azure CLI in a virtual environment is the right path. Detailed instructions for installing with `pip` is provided in https://github.com/Azure/azure-cli/issues/20476.

> we found the release cycle for az-cli differs than the full tool, so bugs aren't always resolved at the same time.

`az-cli` is not the official Azure CLI. Azure CLI's package on PyPI is `azure-cli`.
